File: p004.hs
{- 
 - Solution to Project Euler problem 4
 - By Nayuki Minase
 - 
 - http://nayuki.eigenstate.org/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}


reverseInt = revInt 0 where
	revInt acc 0 = acc
	revInt acc n = revInt (acc * 10 + (mod n 10)) (div n 10)

isPalindrome n = n == (reverseInt n)

ans = foldl1 max [a * b | a <- [100..999], b <- [100..999], isPalindrome (a * b)]
main = putStrLn (show ans)

File: p004.py
# 
# Solution to Project Euler problem 4
# by Project Nayuki
# 
# http://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	max_palindrome = 0
	for i in range(100, 1000):
		for j in range(100, 1000):
			k = i * j
			s = str(k)
			if s == s[::-1] and k > max_palindrome:
				max_palindrome = k
	return str(max_palindrome)


if __name__ == "__main__":
	print(compute())
